File: Atividade 4.2.py
def transfer(data: int, gen: int):
  odata = data.bit_length() - 1
  orgen = gen.bit_length() - 1
  order = odata + orgen
  send = rest = data << orgen
  for i in range(odata + 1):
    if rest & (1 << (order - i)):
      rest ^= gen << (odata - i)
  return send ^ rest

def receiver(data: int, gen: int):
  odata = data.bit_length() - 1
  orgen = gen.bit_length() - 1
  order = odata - orgen
  for i in range(order + 1):
    if data & (1 << (odata - i)):
      data ^= gen << (order - i)
  return data

def main():
  data, gen = 0b1101011011, 0b10011
  res = transfer(data, gen)
  print(f"Resposta: {bin(res)}")
  res = receiver(res, gen)
  print(f"Checagem: {bin(res)}")

if __name__ == "__main__": main()

File: Atividade 4.2.c
#include <stdio.h>

typedef unsigned long long uint64;
typedef unsigned char uint8;

uint8 order(uint64 data) {
  uint8 bits = 0;
  while (data >>= 1) bits++;
  return bits;
}

uint64 transfer(uint64 data, uint64 gen) {
  uint8 orddt = order(data);
  uint8 ordgn = order(gen);
  uint8 sumord = orddt + ordgn;
  uint64 send = data << ordgn;
  uint64 temp = send;
  for (uint8 i = 0; i <= orddt; i++)
    if (temp & (1ull << (sumord - i)))
      temp ^= gen << (orddt - i);
  return send ^ temp;
}

uint64 receiver(uint64 data, uint64 gen) {
  uint8 orddt = order(data);
  uint8 ordgn = order(gen);
  uint8 sumord = orddt + ordgn;
  for (uint8 i = 0; i <= orddt; i++)
    if (data & (1ull << (sumord - i)))
      data ^= gen << (orddt - i);
  return data;
}

void main() {
  uint64 gen = 0b10011;
  uint64 data = 0b1101011011;
  uint64 res1 = transfer(data, gen);
  printf("Resposta: %llu\n", res1);
  uint64 res2 = receiver(res1, gen);
  printf("Checagem: %llu\n", res2);
}